[hwasan] Update (Post-)DominatorTreeAnalysis and LoopAnalysis incrementally

thurstond opened the pull request:
HWAddressSanitizerPass::run sanitizes functions one by one. The sanitization of each function - which may split blocks via insertShadowTagCheck - may result in some cached analyses are invalid. This matters because sanitizeFunction(F', FAM) may indirectly call the global stack safety analysis, hence we need to make sure the analyses of F are up to date.

Bug report: https://github.com/llvm/llvm-project/issues/66934

thurstond committed:
[hwasan] Invalidate DominatorTreeAnalysis after each function is sanitized

HWAddressSanitizerPass::run sanitizes functions one by one. The sanitization of each function - which may split blocks via insertShadowTagCheck - may result in some cached analyses are invalid. This matters because sanitizeFunction(F', FAM) may indirectly call the global stack safety analysis, hence we need to make sure the analyses of F are up to date.

Bug report: https://github.com/llvm/llvm-project/issues/66934

fhahn commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -430,8 +430,16 @@ PreservedAnalyses HWAddressSanitizerPass::run(Module &M,
+    // some cached analyses are invalid. This matters because
+    // sanitizeFunction(F', FAM) may indirectly call the global stack safety
+    // analysis, hence we need to make sure the analyses of F are up to date.
+    PreservedAnalyses PA = PreservedAnalyses::all();
```
This is confusing, in particular because below it marks all analysis as not preserved. Should we invalidate all analysis here?

Also, the dominator tree should be very easy to preserve, to save some compile-time: `SplitBlockAndInsertIfThen` takes an optional DomTreeUpdater object to update the dominator tree.

I didn't take a look at how the pass uses the dominator tree, but if the dominator tree isn't updated after making changes to the CFG and is later used again in the same function, then it may use an out-of-date dominator tree.

thurstond replied:
> This is confusing, in particular because below it marks all analysis as not preserved. Should we invalidate all analysis here?

ASan, MSan, DFSan and HWASan all share the same "invalidate all the analyses, including GlobalsAA, at the end of the pass" boilerplate code, so perhaps it's not strictly necessary for HWASan.

But even if all the analyses are stale at the end of the HWASan pass, they don't necessarily need to be eagerly invalidated after each function is sanitized, as long as the analysis is not used by the rest of the HWASan pass. The stack safety global analysis is a special case because, by definition, it looks at every function.

> Also, the dominator tree should be very easy to preserve, to save some compile-time: `SplitBlockAndInsertIfThen` takes an optional DomTreeUpdater object to update the dominator tree.

That's a good idea, thanks! I'll update the pull request shortly.

thurstond replied:
P.S. I'm honored to have two Florians review my patch :grinning:

thurstond committed:
[hwasan] Update DominatorTreeAnalysis and LoopAnalysis incrementally

This patch incrementally updates the DominatorTreeAnalysis and
LoopAnalysis whenever SplitBlockAndInsertIfThen is called, which fixes
the following issue:
  Suppose we have two functions, F and G. HWAddressSanitizerPass::run
  will first sanitize F, which potentially makes the analyses of F out of
  date. When G is then sanitized, it may call the global stack safety
  analysis, which can crash if the analysis of F is stale.

Additionally, since the DominatorTreeAnalysis and LoopAnalysis are now
incrementally updated, they do not need to be abandoned at the end of
the pass.

Bug report: https://github.com/llvm/llvm-project/issues/66934

fmayer commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -1542,8 +1557,13 @@ void HWAddressSanitizer::sanitizeFunction(Function &F,
+  DominatorTree &DT = FAM.getResult<DominatorTreeAnalysis>(F);
+  PostDominatorTree &PDT = FAM.getResult<PostDominatorTreeAnalysis>(F);
+  DomTreeUpdater DTU(DT, PDT, DomTreeUpdater::UpdateStrategy::Lazy);
+  LoopInfo &LI = FAM.getResult<LoopAnalysis>(F);
```
We could pull the definitions for DT PDT and LI out of the if in 1538 then we need not repeat them.

thurstond replied:
Done, thanks!

thurstond committed:
[hwasan] Update DominatorTreeAnalysis and LoopAnalysis incrementally

This patch incrementally updates the DominatorTreeAnalysis and
LoopAnalysis whenever SplitBlockAndInsertIfThen is called, which fixes
the following issue:
  Suppose we have two functions, F and G. HWAddressSanitizerPass::run
  will first sanitize F, which potentially makes the analyses of F out of
  date. When G is then sanitized, it may call the global stack safety
  analysis, which can crash if the analysis of F is stale.

Additionally, since the DominatorTreeAnalysis and LoopAnalysis are now
incrementally updated, they do not need to be abandoned at the end of
the pass.

Bug report: https://github.com/llvm/llvm-project/issues/66934

fmayer commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -434,6 +438,10 @@ PreservedAnalyses HWAddressSanitizerPass::run(Module &M,
+  // DominatorTreeAnalysis and LoopAnalysis are incrementally updated
+  // throughout this pass whenever SplitBlockAndInsertIfThen is called.
+  PA.preserve<DominatorTreeAnalysis>();
+  PA.preserve<LoopAnalysis>();
```
Can we add PostDominatorTreeAnalysis here too? Seems like DomTreeUpdater also takes care of that.

fmayer replied:
(And if not, please add a comment in the code on why not, because I think that's a natural question)

thurstond replied:
Good point, added.

thurstond committed:
[hwasan] Update (Post-)DominatorTreeAnalysis and LoopAnalysis incrementally

This patch incrementally updates the DominatorTreeAnalysis,
PostDominatorTreeAnalysis and LoopAnalysis whenever
SplitBlockAndInsertIfThen is called, which fixes the following issue:
  Suppose we have two functions, F and G. HWAddressSanitizerPass::run
  will first sanitize F, which potentially makes the analyses of F out of
  date. When G is then sanitized, it may call the global stack safety
  analysis, which can crash if the analysis of F is stale.

Additionally, since the DominatorTreeAnalysis, PostDominatorTreeAnalysis
and LoopAnalysis are now incrementally updated, they do not need to be
abandoned at the end of the pass.

Bug report: https://github.com/llvm/llvm-project/issues/66934

fmayer reviewed (approved):
I give this CL 1 / 2 Florians.

fhahn commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -1520,10 +1537,10 @@ void HWAddressSanitizer::sanitizeFunction(Function &F,
                    Mapping.WithFrameRecord &&
                    !SInfo.AllocasToInstrument.empty());
 
+  DominatorTree &DT = FAM.getResult<DominatorTreeAnalysis>(F);
```
This now may request the DT/LI/PDT in cases where it wasn't before, which should probably avoided. In cases they are only needed for updating them/preserving them, you can use `FAM.getCachedResult`, which won't cause the DT to be computed if it isn't already cached.

fhahn commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -861,8 +871,8 @@ int64_t HWAddressSanitizer::getAccessInfo(bool IsWrite,
-HWAddressSanitizer::insertShadowTagCheck(Value *Ptr,
-                                         Instruction *InsertBefore) {
+HWAddressSanitizer::insertShadowTagCheck(Value *Ptr, Instruction *InsertBefore,
+                                         DomTreeUpdater *DTU, LoopInfo *LI) {
```
Across all those interfaces, both `DTU` and `LI` will always be non-null. Better to pass as reference?

thurstond replied:
Good idea, done!

thurstond committed:
[hwasan] Update (Post-)DominatorTreeAnalysis and LoopAnalysis incrementally

This patch incrementally updates the DominatorTreeAnalysis,
PostDominatorTreeAnalysis and LoopAnalysis whenever
SplitBlockAndInsertIfThen is called, which fixes the following issue:
    Suppose we have two functions, F and G. HWAddressSanitizerPass::run
    will first sanitize F, which potentially makes the analyses of F out of
    date. When G is then sanitized, it may call the global stack safety
    analysis, which can crash if the analysis of F is stale.

Additionally, since the DominatorTreeAnalysis, PostDominatorTreeAnalysis
and LoopAnalysis are now incrementally updated, they do not need to be
abandoned at the end of the pass.

Bug report: https://github.com/llvm/llvm-project/issues/66934

aeubanks commented:
title needs to be updated

fhahn reviewed (approved):
Thanks for the latest updates, LGTM!

thurstond commented:
> title needs to be updated

@aeubanks Done, thanks for the heads up!

thurstond commented:
> Thanks for the latest updates, LGTM!

@fhahn Thanks for all the helpful guidance!
